### DISCUSSION TITLE
Behavior change for routeChangeComplete on React 18?

### DISCUSSION THREAD

**Discussion by gearboxbuilt** [vercel/next.js#40064] [Help]
Have any updates been made for `routeChangeComplete` that may be affected in React 18?

Recently I upgraded a client's website from `react@17.0.2` to `react@18.0.2` and `next@12.1.0` to `12.2.4`. I just received notice from my client today that Google Tag Manager wasn't tracking data. 

Upon investigation, the issue was that the custom `pageview` event wasn't firing on app load but does trigger as expected when a user navigates around the site. 

Here is our pretty standard implementation: 

```
import { useEffect } from 'react'
import { useRouter } from 'next/router'

const { GOOGLE_TAG_TRACKING_ID } = process.env

export const pageview = (url) => {
  window.dataLayer.push({
    event: 'pageview',
    page: url,
  })
}

export default function useGoogleTagManager() {
  const router = useRouter()

  if (!GOOGLE_TAG_TRACKING_ID) return null

  useEffect(() => {
    router.events.on('routeChangeComplete', pageview)
    return () => {
      router.events.off('routeChangeComplete', pageview)
    }
  }, [router.events])

  return null
}
```

Manually firing `pageview` on mount is simple enough and this does solves the issue for my client and me. 

However, I was certain this wasn't the case prior to the update.  For a test, I revert the repo to the previous versions, and sure enough, the  `routeChangeComplete` appears to fire on app mount and this is not the case once the repo is upgraded. 

Has something been updated that I missing?

**Comment by icyJoseph** [comment]
Mmm are you sure that's the only change? 

I am gonna be extremely careful here, but AFAIK, when you load the page, you have a `script`-like tag that invokes the analytics tool, and that counts as the page being visited. 

Because of SPA mechanics, we implement a page view function to track any further events produced by the SPA transitions, which do not reload the `script`-like tag.

I don't think I've ever seen `routeChangeComplete` trigger on application load state. I've tried with a few projects, by reverting to older version with React 17 and Next 12.1.x, and I cannot see what you describe. 

Perhaps there's more than meets the eye? I wanna see if we can get to the bottom of this.

**Comment by ridho-aru** [comment]
perhaps what you need is something like this

```JavaScript
export default function useGoogleTagManager() {
  const router = useRouter()

  useEffect(() => {
      // also, it is not good to have condition before hooks, so try this instead
      if (!GOOGLE_TAG_TRACKING_ID) return;

      // this will trigger pageview on mount (or first load)
      if(typeof window === 'undefined') return;
      
      pageview(window.location.href);
      // or get from the router.asPath (but this needs to fill the array deps with router.asPath)
      // and that will trigger if `asPath` changes also
  }, [])
  
  useEffect(() => {
      // the same as above
      if (!GOOGLE_TAG_TRACKING_ID) return;
  
      router.events.on('routeChangeComplete', pageview)
      return () => {
          router.events.off('routeChangeComplete', pageview)
      }
  }, [router.events])

  return null
}
```

the nature of `routeChangeComplete` won't trigger on app first load, so maybe, I was wondering too if we could get another router event like `pageLoaded` or something similar
